File: 155 Min Stack/min_stack.java
public class MinStack {
    private Stack<Integer> dat, aux;
    private int min;
    /** initialize your data structure here. */
    public MinStack() {
        dat = new Stack<>();
        aux = new Stack<>();
        min = Integer.MAX_VALUE;
    }
    
    public void push(int x) {
        dat.push(x);
        min = Math.min(x, min);
        aux.push(min);
    }
    
    public void pop() {
        if (dat.isEmpty()) return;
        dat.pop();
        aux.pop();
        if (aux.isEmpty()) min = Integer.MAX_VALUE;
        else min = aux.peek();
    }
    
    public int top() {
        if (dat.isEmpty()) return 0;
        return dat.peek();
    }
    
    public int getMin() {
        if (aux.isEmpty()) return 0;
        return aux.peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 * Author: XC
 */

File: 155 Min Stack/min_stack.cpp
class MinStack {
public:
    std::stack<int> st1, st2;
    void push(int x) {
        st1.push(x);
        if (x <= getMin()) st2.push(x);
    }

    void pop() {
        if (st1.top() == getMin()) st2.pop();
        st1.pop();
    }

    int top() {
        return st1.top();
    }

    int getMin() {
        if (st2.empty()) return INT_MAX;
        return st2.top();
    }
}; // Author: XC
